File: greedy/javascript/minimum_rooms.js
/**
 * @param {number[]} start
 * @param {number[]} end
 * @param {number} n
 * @returns {number}
*/
var minimumRooms = function(start, end, n){
	start.sort((a, b) => a - b);
	end.sort((a, b) => a - b);
	
	let i = 0, j = 0, ans = 0, rooms = 0;
	
	// allocate rooms as need them, and keep the maximum we needed
	while (i < n || j < n) {
		if(start[i] <= end[j]) {
			i++;
			rooms++;
		}
		else{
			j++;
			rooms--;
		}
		
		ans = Math.max(ans, rooms);
	}
	
	return ans;
}

start = [900, 940, 950, 1100, 1500, 1800]
end = [910, 1200, 1120, 1130, 1900, 2000]
console.log(minimumRooms(start, end, 6))

start = [900, 1100, 1235]
end = [1000, 1200, 1240]
console.log(minimumRooms(start, end, 3))

File: greedy/cpp/minimum_rooms.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
    public:
    int minimumRooms(int start[], int end[], int n)
    {
    	vector<pair<int, string>> transactions;
    	
    	// make a vector ordered by time containing all transactions
    	for(int i = 0; i < n; i++){
    	    transactions.push_back({start[i], "start"});
    	    transactions.push_back({end[i], "end"});
    	}
    	
    	sort(transactions.begin(), transactions.end());
    	
    	// allocate rooms as need them, and keep the maximum we needed
    	int rooms = 0, ans = 0;
    	for(int i = 0; i < transactions.size(); i++){
    	    if(transactions[i].second == "start") {
    	        rooms++;
    	    }
    	    else if(transactions[i].second == "end"){
    	        rooms--;
    	    }
    	    ans = max(ans, rooms);
    	}
    	

    	return ans;
    }
};

int main() {
    int n;
	cin >> n;

	int start[n], end[n];
	for(int i, a, b; i < n; i++) {
		cin >> start[i] >> end[i];
	}

	Solution solve;
	cout << solve.minimumRooms(start, end, n) << endl;
}